clients/go/pkg/zbc/client.go

```diff
@@ -83,72 +85,90 @@ func (e Error) Error() string {
 	return string(e)
 }
 
+func withStopOnPermanentError(shouldRetryRequest func(ctx context.Context, err error) bool) func(ctx context.Context, err error) bool {
+	return func(ctx context.Context, err error) bool {
+		switch {
+		case errors.Is(err, context.Canceled):
+			return false
+		case errors.Is(err, context.DeadlineExceeded):
+			return false
+		}
+
+		switch status.Code(err) {
+		case codes.Canceled, codes.DeadlineExceeded, codes.InvalidArgument, codes.Unimplemented:
+			return false
+		}
+
+		return shouldRetryRequest(ctx, err)
+	}
+}
+
 func (c *ClientImpl) NewTopologyCommand() *commands.TopologyCommand {
-	return commands.NewTopologyCommand(c.gateway, c.credentialsProvider.ShouldRetryRequest)
+	return commands.NewTopologyCommand(c.gateway, withStopOnPermanentError(c.credentialsProvider.ShouldRetryRequest))
```

Comment:
👍🏼 I like this chaining.
🔧 I was just wondering whether we also should use the same naming, shouldRetryRequest.

Reply:
Done with the client implementation defining its own `shouldRetryRequest`.

```go
func (c *ClientImpl) NewTopologyCommand() *commands.TopologyCommand {
	return commands.NewTopologyCommand(c.gateway, c.shouldRetryRequest)
}
```